teamcity-rest-client-api/src/main/kotlin/org/jetbrains/teamcity/rest/api.kt

```diff
@@ -77,41 +131,95 @@ interface BuildLocator {
     fun pinnedOnly(): BuildLocator
 
     fun limitResults(count: Int): BuildLocator
-    
+
     fun sinceDate(date: Date) : BuildLocator
 
     fun latest(): Build?
+    fun all(): Sequence<Build>
+
+    @Deprecated(message = "use all() which returns lazy sequence",
+                replaceWith = ReplaceWith("all().toList()"))
     fun list(): List<Build>
+    @Deprecated(message = "use includeFailed()",
+                replaceWith = ReplaceWith("includeFailed()"))
+    fun withAnyStatus(): BuildLocator
 }
 
-data class ProjectId(val stringId: String)
+data class ProjectId(val stringId: String) {
+    override fun toString(): String = stringId
+}
 
-data class BuildId(val stringId: String)
+data class BuildId(val stringId: String) {
+    override fun toString(): String = stringId
+}
 
-data class QueuedBuildId(val stringId: String)
+data class TestId(val stringId: String) {
+    override fun toString(): String = stringId
+}
 
-data class ChangeId(val stringId: String)
+data class ChangeId(val stringId: String) {
+    override fun toString(): String = stringId
+}
 
-data class BuildConfigurationId(val stringId: String)
+data class BuildConfigurationId(val stringId: String) {
+    override fun toString(): String = stringId
+}
 
-data class VcsRootId(val stringId: String)
+data class VcsRootId(val stringId: String) {
+    override fun toString(): String = stringId
+}
+
+data class BuildProblemId(val stringId: String) {
+    override fun toString(): String = stringId
+}
+
+data class BuildProblemType(val stringType: String) {
+    override fun toString(): String = stringType
+
+    val isSnapshotDependencyError: Boolean
+        get() = stringType == "SNAPSHOT_DEPENDENCY_ERROR_BUILD_PROCEEDS_TYPE" ||
+                stringType == "SNAPSHOT_DEPENDENCY_ERROR"
+
+    companion object {
+        val FAILED_TESTS = BuildProblemType("TC_FAILED_TESTS")
+    }
+}
 
 interface Project {
     val id: ProjectId
     val name: String
     val archived: Boolean
-    val parentProjectId: ProjectId
+    val parentProjectId: ProjectId?
 
     /**
      * Web UI URL for user, especially useful for error and log messages
      */
+    fun getHomeUrl(branch: String? = null): String
+    fun getTestHomeUrl(testId: TestId): String
+
+    val childProjects: List<Project>
+    val buildConfigurations: List<BuildConfiguration>
+    val parameters: List<Parameter>
+
+    fun setParameter(name: String, value: String)
+
+    fun createVcsRoot(id: VcsRootId, name: String, type: VcsRootType, properties: Map<String, String>): VcsRoot
```

Comment:
This violates [item 2 from the contribution guideline](https://github.com/JetBrains/teamcity-rest-client/blob/master/CONTRIBUTING.md). What should I pass here if I want to create a Git VCS root with a specific fetch URL?
I suggest to add a method `createGitVcsRoot` and pass the main parameters (fetch URL, default branch, branch spec, maybe authentication settings?) as explicit named arguments to it. You can keep `additionalProperties` argument for other properties, and also you can leave this method to support creation of VCS roots for other VCSes.

Reply:
Yeah, and authentication settings, 4 cases of them at least. There isn't a simple configuration of a VCS root. I've added a javadoc.